src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,325 @@
+def is_primitive (ψ : add_char R R') : Prop :=
+∀ (a : R), a ≠ 0 → is_nontrivial (mul_shift ψ a)
+
+/-- The map associating to `a : R` the multiplicative shift of `ψ` by `a`
+is injective when `ψ` is primitive. -/
+lemma to_mul_shift_inj_of_is_primitive {ψ : add_char R R'} (hψ : is_primitive ψ) :
+  function.injective ψ.mul_shift :=
+begin
+  intros a b h,
+  apply_fun (λ x, x * mul_shift ψ (-b)) at h,
+  simp only [mul_shift_mul, mul_shift_zero, add_right_neg] at h,
+  have h₂ := hψ (a + (- b)),
+  rw [h, is_nontrivial_iff_ne_trivial, ← sub_eq_add_neg, sub_ne_zero] at h₂,
+  exact not_not.mp (λ h, h₂ h rfl),
+end
+
+-- `add_comm_group.equiv_direct_sum_zmod_of_fintype`
+-- gives the structure theorem for finite abelian groups.
+-- This could be used to show that the map above is a bijection.
+-- We leave this for a later occasion.
+
+/-- When `R` is a field `F`, then a nontrivial additive character is primitive -/
+lemma is_primitive_of_is_nontrivial {F : Type u} [field F]
+ (ψ : add_char F R') (hψ : is_nontrivial ψ) : is_primitive ψ :=
+begin
+  intros a ha,
+  cases hψ with x h,
+  use (a⁻¹ * x),
+  rwa [mul_shift_apply, to_add_of_add, mul_inv_cancel_left₀ ha],
+end
+
+/-- Structure for a primitive additive character on a finite ring `R` into a cyclotomic extension
+of a field `R'`. It records which cyclotomic extension it is, the character, and the
+fact that the character is primitive. -/
+@[nolint has_inhabited_instance] -- can't prove that they always exist
+structure primitive_add_char (R : Type u) [comm_ring R] [fintype R] (R' : Type v) [field R'] :=
+(n : pnat)
+(char : add_char R (cyclotomic_field n R'))
+(prim : is_primitive char)
+
+/-!
+### Additive characters on `zmod n`
+-/
+
+variables {C : Type v} [comm_ring C]
+
+/-- We can define an additive character on `zmod n` when we have an `n`th root of unity `ζ : C`. -/
+def zmod_char (n : ℕ) [fact (0 < n)] {ζ : C} (hζ : ζ ^ n = 1) : add_char (zmod n) C :=
+{ to_fun := λ (a : multiplicative (zmod n)), ζ ^ a.to_add.val,
+  map_one' := by simp only [to_add_one, zmod.val_zero, pow_zero],
+  map_mul' := λ x y, by rw [to_add_mul, ← pow_add, zmod.val_add (to_add x) (to_add y),
+                            ← pow_eq_pow_mod _ hζ] }
+
+/-- An additive character on `zmod n` is nontrivial iff it takes a value `≠ 1` on `1`. -/
+lemma zmod_char_is_nontrivial_iff (n : ℕ) [fact (0 < n)] (ψ : add_char (zmod n) C) :
+  is_nontrivial ψ ↔ ψ (of_add 1) ≠ 1 :=
+begin
+  refine ⟨_, λ h, ⟨1, h⟩⟩,
+  contrapose!,
+  rintros h₁ ⟨a, ha⟩,
+  have ha₁ : a = a.val • 1,
+  { rw [nsmul_eq_mul, mul_one], exact (zmod.nat_cast_zmod_val a).symm },
+  have ha₂ : of_add a = (of_add 1) ^ a.val := by rw [← of_add_nsmul _ a.val, ← ha₁],
+  rw [ha₂, map_pow, h₁, one_pow] at ha,
+  exact ha rfl,
+end
+
+/-- A primitive additive character on `zmod n` takes the value `1` only at `0`. -/
+lemma primitive_zmod_char_eq_one_iff (n : ℕ) [fact (0 < n)]
+ {ψ : add_char (zmod n) C} (hψ : is_primitive ψ) (a : zmod n) :
```

Comment:
```suggestion
  {ψ : add_char (zmod n) C} (hψ : is_primitive ψ) (a : zmod n) :
```
Also in other places.

Reply:
I've been using one space of indentation for arguments (when more than one line is needed) and two spaces for the statement, to be able to distinguish the two more easily. If this is frowned upon, I can change it.

Reply:
I understand the reasoning, but it differs from what is done in the rest of mathlib.
The current style guide prescribes double spacing everywhere, and this used (almost) uniformly in all of mathlib. So it would be good to stick to that.

Reply:
OK; I've changed the single to double spaces.